Test runner for batched gradient computation with vmap

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43664 Test runner for batched gradient computation with vmap**

This PR implements the test runner for batched gradient computation with
vmap. It also implements the batching rule for sigmoid_backward and
tests that one can compute batched gradients with sigmoid (and batched
2nd gradients).

Test Plan:
- New tests: `python test/test_vmap.py -v`

Differential Revision: [D23358555](https://our.internmc.facebook.com/intern/diff/D23358555)

zou3519 committed:
Test runner for batched gradient computation with vmap

This PR implements the test runner for batched gradient computation with
vmap. It also implements the batching rule for sigmoid_backward and
tests that one can compute batched gradients with sigmoid (and batched
2nd gradients).

Test Plan:
- New tests: `python test/test_vmap.py -v`

[ghstack-poisoned]

zou3519 committed:
Update on "Test runner for batched gradient computation with vmap"


This PR implements the test runner for batched gradient computation with
vmap. It also implements the batching rule for sigmoid_backward and
tests that one can compute batched gradients with sigmoid (and batched
2nd gradients).

Test Plan:
- New tests: `python test/test_vmap.py -v`

[ghstack-poisoned]

zou3519 commented on test/test_vmap.py:
```diff
@@ -640,35 +643,42 @@ def randn(size, device='cpu', dtype=torch.float):
+#
+# check_view: Test if the first returned output is a view of the first input
+# check_propagates_grad: Test if the operation propagates gradients.
+def _vmap_test(self, op, inputs, in_dims=0, out_dims=0,
```
This function was moved line-for-line from TestVmapOperators._vmap_test

zou3519 commented on test/test_vmap.py:
```diff
@@ -1150,5 +1160,100 @@ def test_no_random_op_support(self):
+    # Regression.
+    # It might be useful to have a test that computes batched first gradients and
+    # then uses those to compute batched second gradients in the future.
+    def _batched_grad_grad_test(self, op, args, kwargs, output_process_fn=lambda x: x, batch_size=3):
```
One question is: I could make this (and _batched_grad_grad_test) a part of gradcheck, but I'm not sure if I should.

The nice thing about making it a part of gradcheck is that we can be sure that "batched gradient" works for literally all operators in PyTorch. The downside of making it a part of gradcheck right now is modifying a lot of callsites of gradcheck to not check this (because _batched_grad_grad_test fails for a lot of operators).

I think in the long run we should have gradcheck test for batched gradient, but it might take a while to get there

ezyang replied:
We have a sort of similar situation with complex numbers. My current thinking is that it's fine to not make it part of gradcheck in the beginning when it's only a handful of ops that support batched gradient. Then as you start adding support fo more operators, at some point you will get to the point where you are going for "completion"; at that point, it makes sense to flip the switch, and go from whitelisting ops which do work, to blacklisting ops that don't work.

albanD commented on test/test_vmap.py:
```diff
@@ -640,35 +643,42 @@ def randn(size, device='cpu', dtype=torch.float):
+    if check_view:
+        result_as_tuple = (result,) if op_has_single_return else result
+        for output in result_as_tuple:
+            self.assertEqual(output.data_ptr() - output.storage_offset() * output.element_size(),
```
Note that you could try as well to check that `output._base is inputs[0] if inputs[0]._base is None else inputs[0]._base` ?

zou3519 replied:
I like that better, actually. I'll update the code to do that

albanD commented on test/test_vmap.py:
```diff
@@ -1150,5 +1160,100 @@ def test_no_random_op_support(self):
+
+def differentiable(args):
+    return tuple(arg for arg in as_tuple(args)
+                 if isinstance(arg, torch.Tensor) and arg.requires_grad)
```
I never know which one is supposed to be better `isinstance(arg, torch.Tensor)` or `torch.is_tensor(arg)` ?

zou3519 replied:
https://github.com/pytorch/pytorch/blob/0ab83f7f9fcbbc9b7a709b99b38577d14c22a7b7/torch/__init__.py#L226-L231 We seem to recommend that users use `isinstance(arg, torch.Tensor)`.

zou3519 commented on test/test_vmap.py:
```diff
@@ -1,8 +1,11 @@
+from torch import Tensor, vmap
 import functools
 import warnings
+from torch._six import istuple
```
Good point. I was copy-pastaing some code

zou3519 committed:
Update on "Test runner for batched gradient computation with vmap"


This PR implements the test runner for batched gradient computation with
vmap. It also implements the batching rule for sigmoid_backward and
tests that one can compute batched gradients with sigmoid (and batched
2nd gradients).

Test Plan:
- New tests: `python test/test_vmap.py -v`

Differential Revision: [D23358555](https://our.internmc.facebook.com/intern/diff/D23358555)

[ghstack-poisoned]

ezyang commented on test/test_vmap.py:
```diff
@@ -1150,5 +1159,100 @@ def test_no_random_op_support(self):
+    elif isinstance(x, list):
+        return tuple(x)
+    else:
+        return x,
```
No action really necessary here, but we have some other code like this in other places. It reminds me a bit of https://jax.readthedocs.io/en/latest/pytrees.html too

ezyang commented on test/test_vmap.py:
```diff
@@ -1150,5 +1159,100 @@ def test_no_random_op_support(self):
+    def test_sigmoid(self, device):
+        # Maybe we can make the "check that the slow fallback was not invoked"
+        # into a context manager, because it's used a lot. I'll leave that for
+        # future work.
```
Yes, the Not regex here looks very brittle; you won't notice it's out of date if you change the message

zou3519 replied:
Yup, I made it more structured in another PR (https://github.com/pytorch/pytorch/pull/43731)

ezyang commented on test/test_vmap.py:
```diff
@@ -1150,5 +1159,100 @@ def test_no_random_op_support(self):
+    #       that should be differentiated.
+    # batch_size: the batch dim size for the batched grad
+    #
+    # NB: we only test computing batched gradients are in the second gradient
```
grammar problem here

ezyang commented on test/test_vmap.py:
```diff
@@ -1150,5 +1159,100 @@ def test_no_random_op_support(self):
+    # NB: we only test computing batched gradients are in the second gradient
+    # computation. One specific use case that does this is computing the hessian
+    # matrix of a scalar-valued function; this is useful in Bayesian Logistic
+    # Regression.
```
If you've got a citation, that would be a great thing to put here

zou3519 replied:
No citation, I am just quoting something from someone

ezyang commented on test/test_vmap.py:
```diff
@@ -1150,5 +1159,100 @@ def test_no_random_op_support(self):
+                                          create_graph=True)
+        first_grads = differentiable(first_grads)
+        if len(first_grads) == 0:
+            return  # None of the grads actually depend on the input
```
Maybe this should just fail here? It implies that you shouldn't have run this particular test on the operator in question--seems more likely to be a bug than intentional

zou3519 replied:
yeah, that sounds reasonable

ezyang reviewed (approved):
Sure. I cc'ed @mruberry in case he had more testing related thoughts

zou3519 committed:
Update on "Test runner for batched gradient computation with vmap"


This PR implements the test runner for batched gradient computation with
vmap. It also implements the batching rule for sigmoid_backward and
tests that one can compute batched gradients with sigmoid (and batched
2nd gradients).

Test Plan:
- New tests: `python test/test_vmap.py -v`

Differential Revision: [D23358555](https://our.internmc.facebook.com/intern/diff/D23358555)

[ghstack-poisoned]

mruberry reviewed (approved):
Hey @zou3519, I wish I was more familiar with your vmap work (I hope to become more familiar with it soon!), but just from a test perspective this looks pretty good.
